File: leetcode/963.cc
#include <vector>
#include <limits>
#include <cmath>


using namespace std;


class Solution {
  private:
    vector<vector<int>> points;

  public:
    long long CalLengthSquare(int a, int b) {
        auto& pa = points[a];
        auto& pb = points[b];
        int x_diff = pa[0] - pb[0];
        int y_diff = pa[1] - pb[1];
        return 1LL * x_diff * x_diff + y_diff * y_diff;
    }

    bool CheckAngle(int a, int b, int c) {
        return CalLengthSquare(a, b) + CalLengthSquare(b, c) == CalLengthSquare(a, c);
    }

    bool IsRectangle(int a, int b, int c, int d) {
        return CheckAngle(a, b, c) && CheckAngle(b, c, d) && CheckAngle(c, d, a) && CheckAngle(d, a, b);
    }

    double CalSize(int a, int b, int c, int d) {
        return sqrt(CalLengthSquare(a, b)) * sqrt(CalLengthSquare(b, c));
    }

    double minAreaFreeRect(vector<vector<int>>& points) {
        this->points = points;
        double result = numeric_limits<double>::max();
        int n = points.size();
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (j == i)
                    continue;
                for (int k = 0; k < n; ++k) {
                    if (k == i || k == j)
                        continue;
                    for (int w = 0; w < n; ++w) {
                        if (w == i || w == j || w == k)
                            continue;

                        if (!IsRectangle(i, j, k, w))
                            continue;
                        result = min(result, CalSize(i, j, k, w));
                    }
                }
            }
        }

        if (result == numeric_limits<double>::max())
            return 0;
        else
            return result;
    }
};

File: leetcode/963.py
from collections import defaultdict
import math


class Solution:
    def get_length_square(self, p1, p2):
        diff_x, diff_y = p1[0] - p2[0], p1[1] - p2[1]
        return diff_x * diff_x + diff_y * diff_y

    def minAreaFreeRect(self, points: List[List[int]]) -> float:
        n = len(points)
        if n < 4:
            return 0

        point_groups = defaultdict(list)
        for i in range(n):
            pi = points[i]
            for j in range(i+1, n):
                pj = points[j]
                mid_x, mid_y = (pi[0] + pj[0]) / 2, (pi[1] + pj[1]) / 2
                len_square = self.get_length_square(pi, pj)
                point_groups[(mid_x, mid_y, len_square)].append((pi, pj))

        result = float('inf')
        for group in point_groups.values():
            g = len(group)
            if g <= 1:
                continue
            for i in range(g):
                p1, _ = group[i]
                for j in range(i+1, g):
                    p3, p4 = group[j]
                    area = math.sqrt(self.get_length_square(p1, p3)) * math.sqrt(self.get_length_square(p1, p4))
                    result = min(result, area)

        if result < float('inf'):
            return result
        else:
            return 0
